twofactor_privacyidea/lib/Provider/TwoFactorPrivacyIDEAProvider.php

```diff
@@ -36,452 +39,570 @@
+        {
             // We have not gotten any authentication information whatsoever. This code should never be reached, if the
             // client is sane.
-            $error_message = $this->trans->t("Failed to authenticate.");
+            $errorMessage = $this->trans->t("Failed to authenticate.");
             $errorCode = 1;
             $this->log("info", "User failed to authenticate. No OTP value.");
         }
-        if (class_exists('OCP\Authentication\TwoFactorAuth\TwoFactorException')) {
+
+        if (class_exists('OCP\Authentication\TwoFactorAuth\TwoFactorException'))
+        {
             // This is the behaviour for OC >= 9.2
-            throw new TwoFactorException($error_message, $errorCode);
-        } else {
+            throw new TwoFactorException($errorMessage, $errorCode);
+        }
+        else
+        {
             // This is the behaviour for OC == 9.1 and NC.
             return false;
         }
     }
 
-    public function getClientIP()
+    /**
+     * Ask privacyIDEA to trigger all challenges for a given username via
+     * the /validate/triggerchallenge API request.
+     * If the request was successful, return a list of messages from privacyIDEA's response.
+     * If the request failed with HTTP Code 400 and passOnNoToken is activated, pi_no_auth_required is set to indicate that the
+     * authentication can be skipped. Returns an info string in this case that is displayed instead of the message.
+     * Otherwise, if the request failed for any reason, a TriggerChallengeException is raised.
+     *
+     * @param string $username user for which privacyIDEA should trigger challenges
+     * @return string
+     * @throws AdminAuthException
+     * @throws Exception
+     */
+    private function triggerChallenge(string $username): string
     {
-        if (array_key_exists('HTTP_X_FORWARDED_FOR', $_SERVER)) {
-            return $_SERVER["HTTP_X_FORWARDED_FOR"];
-        } else if (array_key_exists('REMOTE_ADDR', $_SERVER)) {
-            return $_SERVER["REMOTE_ADDR"];
-        } else if (array_key_exists('HTTP_CLIENT_IP', $_SERVER)) {
-            return $_SERVER["HTTP_CLIENT_IP"];
+        $this->session->set("pi_hideOTPField", true);
+        $url = $this->getBaseUrl() . "validate/triggerchallenge";
+        $options = $this->getClientOptions();
+        $adminUser = $this->getAppValue('serviceaccount_user', '');
+        $adminPassword = $this->getAppValue('serviceaccount_password', '');
+        $realm = $this->getAppValue('realm', '');
+        $token = $this->getAuthToken($adminUser, $adminPassword);
+
+        $this->session->set("pi_authorization", $token);
+
+        $options["body"] = ["user" => $username, "realm" => $realm];
+        $options["headers"] = ["PI-Authorization" => $token];
+
+        try
+        {
+            $client = $this->httpClientService->newClient();
+            $result = $client->post($url, $options);
+            $ret = $this->processPIResponse($result);
```

Comment:
only put the stuff in the try block that causes the caught exception (e.g. $ret = $this->processPIResponse($result);)

Reply:
right. thanks.